## Association Rule

### Import Libraries

In [4]:
import pandas as pd
import numpy as np
#!pip install mlxtend
#from mlextend.frequent_patterns import apriori,association_rules
#from mlextend.preprocessing import TransactionEncoder

In [6]:
!pip install mlxtend

In [7]:
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [11]:
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

In [12]:
titanic=pd.read_csv("Titanic.csv")
titanic.head()

Class Gender    Age Survived
0   3rd   Male  Child       No
1   3rd   Male  Child       No
2   3rd   Male  Child       No
3   3rd   Male  Child       No
4   3rd   Male  Child       No

### Pre-Processing

In [13]:
df=pd.get_dummies(titanic) #converting all the categorical into numerical columns
df.head()

Class_1st  Class_2nd  Class_3rd  Class_Crew  Gender_Female  Gender_Male  \
0          0          0          1           0              0            1   
1          0          0          1           0              0            1   
2          0          0          1           0              0            1   
3          0          0          1           0              0            1   
4          0          0          1           0              0            1   

   Age_Adult  Age_Child  Survived_No  Survived_Yes  
0          0          1            1             0  
1          0          1            1             0  
2          0          1            1             0  
3          0          1            1             0  
4          0          1            1             0

#### Under the class column, we have 4 values: 1st,2nd,3rd, and crew. So 4 columns have been created accordingly for "class" column. Similarly we had 2 values under "gender" column which is male, female. So 2 columns have been created accordingly for "gender" column. Similarly, its been done for "Age" & "Survived" column.

### Apriori Algorithm

#### In the above o/p, we have totally 10 columns. So among that 10 columns we have to identify the frequent item sets.

#### Calculating the frequent item sets :

In [16]:
frequent_itemsets = apriori(df,min_support=0.1,use_colnames=True)
frequent_itemsets

support                                           itemsets
0   0.147660                                        (Class_1st)
1   0.129487                                        (Class_2nd)
2   0.320763                                        (Class_3rd)
3   0.402090                                       (Class_Crew)
4   0.213539                                    (Gender_Female)
5   0.786461                                      (Gender_Male)
6   0.950477                                        (Age_Adult)
7   0.676965                                      (Survived_No)
8   0.323035                                     (Survived_Yes)
9   0.144934                             (Age_Adult, Class_1st)
10  0.118582                             (Age_Adult, Class_2nd)
11  0.231713                           (Class_3rd, Gender_Male)
12  0.284871                             (Age_Adult, Class_3rd)
13  0.239891                           (Class_3rd, Survived_No)
14  0.391640                          (Class_Crew, Gender_Male)
15  0.402090                            (Age_Adult, Class_Crew)
16  0.305770                          (Class_Crew, Survived_No)
17  0.193094                         (Age_Adult, Gender_Female)
18  0.156293                      (Gender_Female, Survived_Yes)
19  0.757383                           (Age_Adult, Gender_Male)
20  0.619718                         (Gender_Male, Survived_No)
21  0.166742                        (Survived_Yes, Gender_Male)
22  0.653339                           (Age_Adult, Survived_No)
23  0.297138                          (Age_Adult, Survived_Yes)
24  0.209905                (Age_Adult, Class_3rd, Gender_Male)
25  0.191731              (Gender_Male, Class_3rd, Survived_No)
26  0.216265                (Age_Adult, Class_3rd, Survived_No)
27  0.391640               (Age_Adult, Class_Crew, Gender_Male)
28  0.304407             (Gender_Male, Class_Crew, Survived_No)
29  0.305770               (Age_Adult, Class_Crew, Survived_No)
30  0.143571           (Age_Adult, Gender_Female, Survived_Yes)
31  0.603816              (Age_Adult, Gender_Male, Survived_No)
32  0.153567             (Age_Adult, Survived_Yes, Gender_Male)
33  0.175829   (Age_Adult, Gender_Male, Class_3rd, Survived_No)
34  0.304407  (Age_Adult, Gender_Male, Class_Crew, Survived_No)

#### Displaying the records whose lift_ratio is greater than 0.7 :

In [21]:
rules=association_rules(frequent_itemsets,metric="lift",min_threshold=0.7)
rules

antecedents                             consequents  \
0                  (Age_Adult)                             (Class_1st)   
1                  (Class_1st)                             (Age_Adult)   
2                  (Age_Adult)                             (Class_2nd)   
3                  (Class_2nd)                             (Age_Adult)   
4                  (Class_3rd)                           (Gender_Male)   
..                         ...                                     ...   
101  (Class_Crew, Survived_No)                (Age_Adult, Gender_Male)   
102                (Age_Adult)  (Survived_No, Class_Crew, Gender_Male)   
103              (Gender_Male)    (Age_Adult, Class_Crew, Survived_No)   
104               (Class_Crew)   (Age_Adult, Survived_No, Gender_Male)   
105              (Survived_No)    (Age_Adult, Class_Crew, Gender_Male)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.950477            0.147660  0.144934    0.152486  1.032680   
1              0.147660            0.950477  0.144934    0.981538  1.032680   
2              0.950477            0.129487  0.118582    0.124761  0.963505   
3              0.129487            0.950477  0.118582    0.915789  0.963505   
4              0.320763            0.786461  0.231713    0.722380  0.918520   
..                  ...                 ...       ...         ...       ...   
101            0.305770            0.757383  0.304407    0.995542  1.314450   
102            0.950477            0.304407  0.304407    0.320268  1.052103   
103            0.786461            0.305770  0.304407    0.387060  1.265851   
104            0.402090            0.603816  0.304407    0.757062  1.253795   
105            0.676965            0.391640  0.304407    0.449664  1.148157   

     leverage  conviction  
0    0.004587    1.005694  
1    0.004587    2.682493  
2   -0.004492    0.994601  
3   -0.004492    0.588085  
4   -0.020555    0.769177  
..        ...         ...  
101  0.072822   54.427079  
102  0.015075    1.023334  
103  0.063931    1.132622  
104  0.061619    1.630802  
105  0.039280    1.105434  

[106 rows x 9 columns]

#### Sorting the above records in descending order based on lift ratio :

In [35]:
rules.sort_values('lift',ascending=False)[0:50] #displaying only the first 50 records

antecedents  \
64              (Age_Adult, Gender_Female)   
69                          (Survived_Yes)   
19                          (Survived_Yes)   
18                         (Gender_Female)   
68                         (Gender_Female)   
65               (Age_Adult, Survived_Yes)   
96                (Age_Adult, Gender_Male)   
101              (Class_Crew, Survived_No)   
47                (Age_Adult, Gender_Male)   
50                            (Class_Crew)   
54               (Class_Crew, Survived_No)   
55                           (Gender_Male)   
103                          (Gender_Male)   
94    (Age_Adult, Class_Crew, Survived_No)   
104                           (Class_Crew)   
93   (Age_Adult, Survived_No, Gender_Male)   
10                            (Class_Crew)   
11                           (Gender_Male)   
51                           (Gender_Male)   
46                 (Age_Adult, Class_Crew)   
78     (Age_Adult, Class_3rd, Gender_Male)   
91                           (Survived_No)   
34                (Class_3rd, Gender_Male)   
39                           (Survived_No)   
56                            (Class_Crew)   
53              (Survived_No, Gender_Male)   
97                 (Age_Adult, Class_Crew)   
100             (Survived_No, Gender_Male)   
98                (Age_Adult, Survived_No)   
99               (Class_Crew, Gender_Male)   
70                (Age_Adult, Gender_Male)   
75                           (Survived_No)   
71                (Age_Adult, Survived_No)   
74                           (Gender_Male)   
22                           (Gender_Male)   
23                           (Survived_No)   
59                (Age_Adult, Survived_No)   
62                            (Class_Crew)   
85                (Class_3rd, Gender_Male)   
84                (Age_Adult, Survived_No)   
57                           (Survived_No)   
92    (Age_Adult, Class_Crew, Gender_Male)   
52               (Class_Crew, Gender_Male)   
105                          (Survived_No)   
58                 (Age_Adult, Class_Crew)   
14                            (Class_Crew)   
15                           (Survived_No)   
63                           (Survived_No)   
45                           (Survived_No)   
40                  (Age_Adult, Class_3rd)   

                               consequents  antecedent support  \
64                          (Survived_Yes)            0.193094   
69              (Age_Adult, Gender_Female)            0.323035   
19                         (Gender_Female)            0.323035   
18                          (Survived_Yes)            0.213539   
68               (Age_Adult, Survived_Yes)            0.213539   
65                         (Gender_Female)            0.297138   
96               (Class_Crew, Survived_No)            0.757383   
101               (Age_Adult, Gender_Male)            0.305770   
47                            (Class_Crew)            0.757383   
50                (Age_Adult, Gender_Male)            0.402090   
54                           (Gender_Male)            0.305770   
55               (Class_Crew, Survived_No)            0.786461   
103   (Age_Adult, Class_Crew, Survived_No)            0.786461   
94                           (Gender_Male)            0.305770   
104  (Age_Adult, Survived_No, Gender_Male)            0.402090   
93                            (Class_Crew)            0.603816   
10                           (Gender_Male)            0.402090   
11                            (Class_Crew)            0.786461   
51                 (Age_Adult, Class_Crew)            0.786461   
46                           (Gender_Male)            0.402090   
78                           (Survived_No)            0.209905   
91     (Age_Adult, Class_3rd, Gender_Male)            0.676965   
34                           (Survived_No)            0.231713   
39                (Class_3rd, Gender_Male)            0.676965   
56              (Survived_No, Gend

#### In the above o/p, we have to identify only the actionable rules, so for that we have to display only those records whose lift ratio is greater than 1. 

#### Displaying those records whose lift ratio is greater than 1 :

In [36]:
rules[rules.lift>1]

antecedents                             consequents  \
0                  (Age_Adult)                             (Class_1st)   
1                  (Class_1st)                             (Age_Adult)   
8                  (Class_3rd)                           (Survived_No)   
9                (Survived_No)                             (Class_3rd)   
10                (Class_Crew)                           (Gender_Male)   
..                         ...                                     ...   
101  (Class_Crew, Survived_No)                (Age_Adult, Gender_Male)   
102                (Age_Adult)  (Survived_No, Class_Crew, Gender_Male)   
103              (Gender_Male)    (Age_Adult, Class_Crew, Survived_No)   
104               (Class_Crew)   (Age_Adult, Survived_No, Gender_Male)   
105              (Survived_No)    (Age_Adult, Class_Crew, Gender_Male)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.950477            0.147660  0.144934    0.152486  1.032680   
1              0.147660            0.950477  0.144934    0.981538  1.032680   
8              0.320763            0.676965  0.239891    0.747875  1.104747   
9              0.676965            0.320763  0.239891    0.354362  1.104747   
10             0.402090            0.786461  0.391640    0.974011  1.238474   
..                  ...                 ...       ...         ...       ...   
101            0.305770            0.757383  0.304407    0.995542  1.314450   
102            0.950477            0.304407  0.304407    0.320268  1.052103   
103            0.786461            0.305770  0.304407    0.387060  1.265851   
104            0.402090            0.603816  0.304407    0.757062  1.253795   
105            0.676965            0.391640  0.304407    0.449664  1.148157   

     leverage  conviction  
0    0.004587    1.005694  
1    0.004587    2.682493  
8    0.022745    1.281251  
9    0.022745    1.052040  
10   0.075412    8.216621  
..        ...         ...  
101  0.072822   54.427079  
102  0.015075    1.023334  
103  0.063931    1.132622  
104  0.061619    1.630802  
105  0.039280    1.105434  

[74 rows x 9 columns]

#### So,out of 106 rows which was there totally (in 21st cell) , now only 74 rows are there whose lift ratio is greater than 1.

#### Displaying those records whose lift ratio is greater than 1 and also whose confidence is greater than 0.75 :

In [37]:
rules[(rules.lift>1) & (rules.confidence>0.75)]

antecedents  \
1                               (Class_1st)   
10                             (Class_Crew)   
13                             (Class_Crew)   
14                             (Class_Crew)   
20                              (Age_Adult)   
21                            (Gender_Male)   
22                            (Gender_Male)   
23                            (Survived_No)   
25                            (Survived_No)   
34                 (Class_3rd, Gender_Male)   
36                 (Class_3rd, Survived_No)   
40                   (Age_Adult, Class_3rd)   
46                  (Age_Adult, Class_Crew)   
48                (Class_Crew, Gender_Male)   
50                             (Class_Crew)   
52                (Class_Crew, Gender_Male)   
54                (Class_Crew, Survived_No)   
56                             (Class_Crew)   
58                  (Age_Adult, Class_Crew)   
60                (Class_Crew, Survived_No)   
62                             (Class_Crew)   
70                 (Age_Adult, Gender_Male)   
71                 (Age_Adult, Survived_No)   
72               (Survived_No, Gender_Male)   
74                            (Gender_Male)   
75                            (Survived_No)   
78      (Age_Adult, Class_3rd, Gender_Male)   
80      (Age_Adult, Class_3rd, Survived_No)   
85                 (Class_3rd, Gender_Male)   
92     (Age_Adult, Class_Crew, Gender_Male)   
94     (Age_Adult, Class_Crew, Survived_No)   
95   (Survived_No, Class_Crew, Gender_Male)   
97                  (Age_Adult, Class_Crew)   
99                (Class_Crew, Gender_Male)   
101               (Class_Crew, Survived_No)   
104                            (Class_Crew)   

                               consequents  antecedent support  \
1                              (Age_Adult)            0.147660   
10                           (Gender_Male)            0.402090   
13                             (Age_Adult)            0.402090   
14                           (Survived_No)            0.402090   
20                           (Gender_Male)            0.950477   
21                             (Age_Adult)            0.786461   
22                           (Survived_No)            0.786461   
23                           (Gender_Male)            0.676965   
25                             (Age_Adult)            0.676965   
34                           (Survived_No)            0.231713   
36                           (Gender_Male)            0.239891   
40                           (Survived_No)            0.284871   
46                           (Gender_Male)            0.402090   
48                             (Age_Adult)            0.391640   
50                (Age_Adult, Gender_Male)            0.402090   
52                           (Survived_No)            0.391640   
54                           (Gender_Male)            0.305770   
56              (Survived_No, Gender_Male)            0.402090   
58                           (Survived_No)            0.402090   
60                             (Age_Adult)            0.305770   
62                (Age_Adult, Survived_No)            0.402090   
70                           (Survived_No)            0.757383   
71                           (Gender_Male)            0.653339   
72                             (Age_Adult)            0.619718   
74                (Age_Adult, Survived_No)            0.786461   
75                (Age_Adult, Gender_Male)            0.676965   
78                           (Survived_No)            0.209905   
80                           (Gender_Male)            0.216265   
85                (Age_Adult, Survived_No)            0.231713   
92                           (Survived_No)            0.391640   
94                           (Gender_Male)            0.305770   
95                             (Age_Adult)            0.304407   
97              (Survived_No, Gender_Male)            0.402090   
99                (Age_Adult, Survived_No)      

#### So,out of 74 rows which was there in the previous o/p, now nearly only 36 rows are there whose lift ratio is greater than 1 and also whose confidence is greater than 0.75. As we could see in 35th cell, most of the lift ratio values falls nearer to the value 1. So, in these kind of cases we can consider another paramter "confidence". Atlast accordingly, we can take the manual business decision.